From ed4531efd6a927f0880e946356afbba0f296a9af Mon Sep 17 00:00:00 2001
From: Craig Kewley <craigx.kewley@intel.com>
Date: Fri, 8 Sep 2017 13:51:45 +0100
Subject: [PATCH 80/96] ASoC: TDF8532: Read error status

Verify the execution of each state change command (start / stop) by
reading the last message error status after each call.

Signed-off-by: Craig Kewley <craigx.kewley@intel.com>
---
 sound/soc/codecs/tdf8532.c | 44 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/sound/soc/codecs/tdf8532.c b/sound/soc/codecs/tdf8532.c
index 8d0a407..f14c72d 100644
--- a/sound/soc/codecs/tdf8532.c
+++ b/sound/soc/codecs/tdf8532.c
@@ -182,6 +182,34 @@ static int tdf8532_get_state(struct tdf8532_priv *dev_data, u8 *state)
 	return ret;
 }
 
+static int tdf8532_dump_dev_error(struct tdf8532_priv *dev_data)
+{
+	u8 error;
+	int ret = 0;
+	char *repl_buff = NULL;
+	struct device *dev = &(dev_data->i2c->dev);
+
+	ret = tdf8532_amp_write(dev_data, GET_ERROR);
+	if (ret < 0)
+		goto out;
+
+	ret = tdf8532_single_read(dev_data, &repl_buff);
+	if (ret < 0)
+		goto out_free;
+
+	error = ((struct get_error_repl *)repl_buff)->error;
+
+	if (error) {
+		dev_err(dev, "%s: 0x%X\n", __func__, error);
+		ret = -EIO;
+	}
+
+out_free:
+	kfree(repl_buff);
+out:
+	return ret;
+}
+
 static int tdf8532_wait_state(struct tdf8532_priv *dev_data, u8 req_state,
 					unsigned long timeout_val)
 {
@@ -217,7 +245,15 @@ static int tdf8532_start_play(struct tdf8532_priv *tdf8532)
 	if (ret < 0)
 		return ret;
 
+	ret = tdf8532_dump_dev_error(tdf8532);
+	if (ret < 0)
+		return ret;
+
 	ret = tdf8532_amp_write(tdf8532, SET_CLK_STATE, CLK_CONNECT);
+	if (ret < 0)
+		return ret;
+
+	ret = tdf8532_dump_dev_error(tdf8532);
 
 	if (ret >= 0)
 		ret = tdf8532_wait_state(tdf8532, STATE_PLAY, ACK_TIMEOUT);
@@ -235,6 +271,10 @@ static int tdf8532_stop_play(struct tdf8532_priv *tdf8532)
 	if (ret < 0)
 		goto out;
 
+	ret = tdf8532_dump_dev_error(tdf8532);
+	if (ret < 0)
+		goto out;
+
 	ret = tdf8532_wait_state(tdf8532, STATE_STBY, ACK_TIMEOUT);
 	if (ret < 0)
 		goto out;
@@ -243,6 +283,10 @@ static int tdf8532_stop_play(struct tdf8532_priv *tdf8532)
 	if (ret < 0)
 		goto out;
 
+	ret = tdf8532_dump_dev_error(tdf8532);
+	if (ret < 0)
+		goto out;
+
 	ret = tdf8532_wait_state(tdf8532, STATE_IDLE, ACK_TIMEOUT);
 
 out:
-- 
1.9.1

